src/main/java/bio/terra/stairway/Flight.java

```diff
@@ -348,6 +336,43 @@ private StepResult stepWithRetry() throws InterruptedException, StairwayExecutio
     return result;
   }
 
+  /**
+   * Helper function to replace the StepResult returned by a step based on {@link FlightDebugInfo}.
+   *
+   * @param initialResult the StepResult initially returned by the step.
+   * @return StepResult to use for the step. May be the initial result.
+   */
+  private StepResult debugStatusReplacement(StepResult initialResult) {
+    if (context().getDebugInfo() == null) {
+      return initialResult;
+    }
+    FlightDebugInfo debugInfo = context().getDebugInfo();
+    // If we are in debug mode and a failure is set for this step AND we have not already
+    // failed here, then insert a failure. We do this right after the step completes but
+    // before the flight logs it so that we can look for dangerous UNDOs.
+    if (debugInfo.getFailAtSteps() != null
+        && context().isDoing()
+        && debugInfo.getFailAtSteps().containsKey(context().getStepIndex())
+        && !debugStepsFailed.contains(context().getStepIndex())) {
+      StepStatus failStatus = debugInfo.getFailAtSteps().get(context().getStepIndex());
+      logger.info(
+          "Failed for debug mode fail step at step {} with result {}",
+          context().getStepIndex(),
+          failStatus);
+      debugStepsFailed.add(context().getStepIndex());
+      return new StepResult(failStatus);
+    }
+    // If we are in debug mode for failing at the last step, and this is the last step, insert a
+    // failure.
+    if (debugInfo.getLastStepFailure()
+        && context().isDoing()
+        && context().getStepIndex() == steps.size() - 1) {
+      logger.info("Failed for debug mode last step failure.");
```

Comment:
For consistency, probably worth adding this step index to debugStepsFailed from above. Shouldn't have an effect now, but in case we ever use that field for something else.

Reply:
Hmmm. As commented, it's specific to failAtSteps and I think it's a little more confusing to explain its dual purpose when that dual purpose isn't relied on. I'm a little inclined to leave as-is and YAGNI let future changes change the field as it's useful for them. Could be persuaded.